pennylane/optimize/rotosolve.py

```diff
@@ -132,340 +212,410 @@ class RotosolveOptimizer:
+                all_keys = set(_nums_frequency) | set(_spectra)
+
+                shape = qml.math.shape(arg)
+                indices = np.ndindex(shape) if len(shape) > 0 else [()]
+                for par_idx in indices:
+                    if par_idx not in all_keys:
+                        raise ValueError(
+                            "Neither the number of frequencies nor the frequency spectrum "
+                            f"was provided for the entry {par_idx} of argument {arg_name}."
+                        )
+
     def step_and_cost(
         self,
         objective_fn,
         *args,
-        num_freqs=None,
-        optimizer=None,
-        optimizer_kwargs=None,
+        nums_frequency=None,
+        spectra=None,
+        shifts=None,
         full_output=False,
         **kwargs,
     ):
         r"""Update args with one step of the optimizer and return the corresponding objective
-        function value prior to the step.
+        function value prior to the step. Each step includes multiple substeps, one per
+        parameter.
 
         Args:
             objective_fn (function): the objective function for optimization. It should take a
                 sequence of the values ``*args`` and a list of the gates ``generators`` as inputs,
                 and return a single value.
-            *args : variable length sequence containing the initial
-                values of the variables to be optimized over or a single float with the initial
-                value.
-            num_freqs (int or array[int]): The number of frequencies in the ``objective_fn`` per
-                parameter. If an ``int``, the same number is used for all parameters; if
-                ``array[int]``, the shape of ``args`` and ``num_freqs`` has to coincide.
-                Defaults to ``num_freqs=1``, corresponding to Pauli rotation gates.
-            optimizer (callable or str): the optimization method used for the univariate
-                minimization if there is more than one frequency with respect to the respective
-                parameter. If a callable, should have the signature
-                ``(fun, **kwargs) -> x_min, y_min``, where ``y_min`` is tracked and returned
-                if ``full_output==True`` but is not relevant to the optimization.
-                If ``"brute"`` or ``"shgo"``, the corresponding global optimizer of SciPy is used.
-                Defaults to ``"brute"``.
-            optimizer_kwargs : keyword arguments for the ``optimizer``. For ``"brute"`` and
-                ``"shgo"``, these kwargs are passed to the respective SciPy implementation.
-                Has to be given as one dictionary, *not variable length*.
+            *args (Sequence): variable length sequence containing the initial values of the
+                variables to be optimized over or a single float with the initial value.
+            nums_frequency (dict[dict]): The number of frequencies in the ``objective_fn`` per
```

Comment:
the change in name is here a breaking change :) So should be added to the 'breaking changes' section of the changelog.